File: 1143.Longest Common Subsequence.py
class Solution(object):
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """
        if len(text1) < len(text2):
            text1, text2 = text2, text1
        print text1, text2
        M, N = len(text1), len(text2)
        dp_common_ended_at = [[0] * N, [0] * N]
        dp_common_range_mx = [[0] * N, [0] * N]
        for m in xrange(M):
            for n in xrange(N):
                if text1[m] == text2[n]:
                    dp_common_ended_at[1][n] = max(
                        n and dp_common_range_mx[1][n - 1],
                        m and dp_common_range_mx[0][n],
                        (m and n and dp_common_range_mx[0][n - 1]) + 1
                    )
                dp_common_range_mx[1][n] = max(
                    m and dp_common_range_mx[0][n],
                    n and dp_common_range_mx[1][n - 1],
                    dp_common_ended_at[1][n]
                )
            dp_common_ended_at[0][:] = [0] * N
            dp_common_ended_at[0], dp_common_ended_at[1] = dp_common_ended_at[1], dp_common_ended_at[0]
            dp_common_range_mx[0][:] = [0] * N
            dp_common_range_mx[0], dp_common_range_mx[1] = dp_common_range_mx[1], dp_common_range_mx[0]
        return dp_common_range_mx[0][-1]

File: 1143.Longest Common Subsequence.rb
# @param {String} text1
# @param {String} text2
# @return {Integer}
def longest_common_subsequence(text1, text2)
    text1, text2 = text2, text1 if text1.size < text2.size
    dp = [0] * (text2.size + 1)
    dp[0] = 0
    for i in (0...text1.size)
        topleft, left = dp[0], 0
        for j in (0...text2.size)
            top = dp[j + 1]
            if text1[i] == text2[j]
                dp[j + 1] = [topleft + 1, left, top].max
            else
                dp[j + 1] = [left, top].max
            end
            topleft, left = top, dp[j + 1]
        end
    end
    dp[-1]
end
